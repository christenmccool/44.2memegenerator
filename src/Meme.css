.Meme {
  height: 300px;
  position: relative;
}

.Meme-img {
  height: 100%;
}

.Meme-remove-div {
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  font-size: 250px;
  color: transparent;
  padding: 0;
}

.Meme-remove-div:hover {
  background-color: grey;
  color: darkgrey;
  opacity: 0.5;
}

.Meme-btn {
  position: absolute;
  left: 0;
  top: 0;
  margin: 3px;
}

.Meme-top-caption, .Meme-bottom-caption {
  position: absolute;
  color: white;
  font-size: 30px;
  text-align: center;
  /* white-space: nowrap; */
}

.Meme-top-caption {
  top: 0;
  left: 50%;
  transform: translateX(-50%);
}

.Meme-bottom-caption {
  bottom: 0;
  left: 50%;
  transform: translateX(-50%);
}