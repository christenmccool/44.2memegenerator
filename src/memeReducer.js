const memesFromStorage = JSON.parse(localStorage.getItem('memes')) || [];

// const INITIAL_STATE = {memes: 
//                         [
//                           {id: 1,
//                             url:"https://www.cheatsheet.com/wp-content/uploads/2020/05/Steve-Carell.jpg",                          
//                             top:"No God!",
//                             bottom: "No please no!"},
//                           {id: 2,
//                             url:"https://i.redd.it/4nodbxs79u311.jpg",
//                             top:"No God!",
//                             bottom: "No please no!"}
//                         ]
//                       };

const INITIAL_STATE = {memes: memesFromStorage};

const memeReducer = (state=INITIAL_STATE, action) => {
  switch (action.type) {
    case "ADD_MEME": {
      localStorage.setItem('memes', JSON.stringify([...state.memes, action.payload]));
      return {...state, 
              memes: [...state.memes, action.payload]};
      }
    case "REMOVE_MEME": {
      localStorage.setItem('memes', JSON.stringify(state.memes.filter(ele => ele.id !== action.id)));
      return {...state, 
              memes: state.memes.filter(ele => ele.id !== action.id)};
      }
    default: return state;
  }
}

export default memeReducer;