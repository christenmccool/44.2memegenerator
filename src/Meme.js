import React from 'react';
import {useDispatch} from 'react-redux';
import './Meme.css';

const Meme = ({id, url, top, bottom}) => {
  const dispatch = useDispatch();
  const handleClick = (evt) => {
    dispatch({type:"REMOVE_MEME", id})
  }

  return (
    <div className="Meme">
      <img className="Meme-img" src={url} alt="meme" />
      {/* <button className="Meme-btn" onClick={handleClick}>X</button>  */}
      <div className="Meme-top-caption">{top}</div>
      <div className="Meme-bottom-caption">{bottom}</div>
      <div className="Meme-remove-div" onClick={handleClick}>X</div>
    </div>
  )
}

export default Meme;