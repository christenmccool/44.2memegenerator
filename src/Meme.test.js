import { render, asFragment } from '@testing-library/react';
import {createStore} from "redux";
import {Provider} from "react-redux";
import memeReducer from './memeReducer';
import Meme from './Meme';

const store = createStore(memeReducer);

test('renders without crashing', () => {
  render(<Provider store={store}><Meme id={1} url="www.google.com" top="hi" bottom="bye"/></Provider>);
  // render(<Meme id={1} url="www.google.com" top="hi" bottom="bye"/>);
  }
);

test('matches snapshot', () => {
  const {asFragment} = render(<Provider store={store}><Meme id={1} url="www.google.com" top="hi" bottom="bye"/></Provider>);
  expect(asFragment()).toMatchSnapshot();
  }
);

test('renders meme', () => {
  const {queryByText, queryByAltText} = render(<Provider store={store}><Meme id={1} url="www.google.com" top="hi" bottom="bye"/></Provider>);
  expect(queryByText('hi')).toBeInTheDocument();
  expect(queryByText('bye')).toBeInTheDocument();
  const image = queryByAltText('a meme');

  expect(image).toBeInTheDocument();

  expect(image).toHaveAttribute('src',
    'www.google.com'
  );
}
);

