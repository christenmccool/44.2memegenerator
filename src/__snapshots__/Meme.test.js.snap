// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Meme"
  >
    <img
      alt="a meme"
      class="Meme-img"
      src="www.google.com"
    />
    <div
      class="Meme-top-caption"
    >
      hi
    </div>
    <div
      class="Meme-bottom-caption"
    >
      bye
    </div>
    <div
      class="Meme-remove-div"
    >
      X
    </div>
  </div>
</DocumentFragment>
`;
